### PR TITLE
Enable batching on inferentia and add tests

### PR DESCRIPTION
**Pull request by jbkyang-nvi**
Enabled batching for inferentia models, added tests for batching and fixed tensorflow bug where first dimension of inputs need to be multiples of num_threads_per_core  * core_per_instance. 
Also added extra error reporting in python backend stub
Testing PR: https://github.com/triton-inference-server/server/pull/3714

### CODE REVIEW

File: inferentia/scripts/gen_triton_model.py [hunk 1]
```diff
@@ -336,7 +342,8 @@ def _validate_output_dict(self, expected_count):
     return init_impl
 
 
-def get_tensorflow_execute_impl():
+def get_tensorflow_execute_impl(enable_batching,
```
Comment (tanmayv25) [inline inferentia/scripts/gen_triton_model.py:345]: Same here.. I don't think you need to pass enable_batching flag here. Just look whether execute function was called with single request or multiple.

**Review by tanmayv25 — CHANGES_REQUESTED** [review]

File: src/pb_stub.cc [hunk 2]
```diff
@@ -494,9 +494,12 @@ Stub::Execute(RequestBatch* request_batch, ResponseBatch* response_batch)
   // If the number of request objects do not match the number of resposne
   // objects throw an error.
   if (response_size != batch_size) {
-    throw PythonBackendException(
+    std::string err =
         "Number of InferenceResponse objects do not match the number of "
-        "InferenceRequest objects.");
+        "InferenceRequest objects. InferenceRequest size is:" +
```
Comment (Tabrizian) [inline src/pb_stub.cc:499]: "InferenceRequest(s) size is"

File: src/pb_stub.cc [hunk 3]
```diff
@@ -494,9 +494,12 @@ Stub::Execute(RequestBatch* request_batch, ResponseBatch* response_batch)
...
+        std::to_string(batch_size) +
+        ", and response size is:" + std::to_string(response_size) + "\n";
```
Comment (Tabrizian) [inline src/pb_stub.cc:501]: "InferenceResponse(s) size is"

File: inferentia/scripts/gen_triton_model.py [hunk 4]
```diff
@@ -628,13 +796,20 @@ def create_model_file(using_tensorflow_model):
     elif FLAGS.model_type == 'pytorch':
         is_tensorflow_model = False
 
+    enable_batching = FLAGS.max_batch_size > 0
```
[the remark this answers is not in the document]
Comment (jbkyang-nvi) [reply inferentia/scripts/gen_triton_model.py:799]: Does `max_batch_size` matter at all if dynamic batching is disabled?

File: inferentia/scripts/gen_triton_model.py (hunk 1 above)
Comment (jbkyang-nvi) [reply inferentia/scripts/gen_triton_model.py:345]: I was trying to speed up execution for the default case (no batching) so there doesn't need to be extra checks

File: inferentia/scripts/gen_triton_model.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (tanmayv25) [reply inferentia/scripts/gen_triton_model.py]: Yes.. The server will reject requests client sends with a batch size greater than `max_batch_size`...
It is like a validation parameter.

File: inferentia/scripts/gen_triton_model.py (hunk 1 above)
Comment (tanmayv25) [reply inferentia/scripts/gen_triton_model.py]: You can speed up the execution when the server sent only a single request to the backend. That is even if the dynamic batching is enabled, the server does not always send multiple requests to the backend. For instance, if the delay time expired and only a single request was received by the server, we don't need to perform extra logic...

File: inferentia/README.md [hunk 5]
```diff
@@ -156,6 +158,24 @@ loaded on cores 2-3. To best engage inferentia device, try setting
 the number of neuron cores to be a proper multiple of the instance
 count.
 
+#### Batching with Pytorch
+
+There are three flags related to batched inputs:
```
Comment (tanmayv25) [inline inferentia/README.md:163]: `gen_triton_model.py`  supports following three options for configuring (triton's dynamic batching)\<link to the triton dynamic batching\>:

1. --preferred_batch_size: <description with link to triton's docs>
2. --max_queue_delay_microseconds: <desc with link to triton's docs>
3. --disable_batch_requests_to_neuron: \<desc\>

File: inferentia/README.md [hunk 6]
```diff
@@ -189,6 +209,10 @@ next available Neuron cores and not specific ones. See
 [Parallel Execution using NEURONCORE_GROUP_SIZES](https://awsdocs-neuron.readthedocs-hosted.com/en/latest/neuron-guide/appnotes/perf/parallel-ncgs.html?highlight=NEURONCORE_GROUP_SIZES)
 for more information.
 
+#### Batching with Tensorflow
+
+The batching flags have the same functionality for Tensorflow
```
Comment (tanmayv25) [inline inferentia/README.md:214]: This is confusing. You should not have separate sections for PyTorch and TensorFlow. Add a common section after 'Serving Inferentia on Triton' section. You can call it `Using Triton's Dynamic Batching` with Inferentia models.

File: inferentia/scripts/gen_triton_model.py [hunk 7]
```diff
@@ -336,7 +351,8 @@ def _validate_output_dict(self, expected_count):
     return init_impl
 
 
-def get_tensorflow_execute_impl():
+def get_tensorflow_execute_impl(enable_dynamic_batching,
```
Comment (tanmayv25) [inline inferentia/scripts/gen_triton_model.py:354]: These python models doesnot do anything to enable triton's batching. When dyanmic batching in triton is enabled then Triton Server will itself provide multiple requests in one execute call if it was able to batch multiple requests. It is inefficient to run the extra logic even if the server was unable to get multiple requests for the python backend. 

enable_dynamic_batching should not appear in python models.. Instead it should rely on the requests list received. If the list contains single request then nothing extra needs to be done.

File: inferentia/scripts/gen_triton_model.py [hunk 8]
```diff
@@ -374,54 +391,133 @@ def execute(self, requests):
...
+            inference_response = pb_utils.InferenceResponse(output_tensors=output_tensors)
+            responses.append(inference_response)
 
         return responses
 '''
     return exec_impl
 
 
-def get_pytorch_execute_impl():
+def get_pytorch_execute_impl(enable_dynamic_batching,
```
Comment (tanmayv25) [inline inferentia/scripts/gen_triton_model.py:519]: See my comment above.. enable_dynamic_batching should not appear here.

**Review by tanmayv25 — CHANGES_REQUESTED** [review]

File: inferentia/scripts/gen_triton_model.py (hunk 7 above)
Comment (jbkyang-nvi) [reply inferentia/scripts/gen_triton_model.py:354]: When `enable_dynamic_batching` is specified though, Triton will send inputs of `[X, input dim]` according to my tests. eg: Input provided from perf_analyzer is [6, 4], model accepts [X, 4] but all the requests sent are of type [1,6,4]. 
This is why `squeeze` is required for batched requests.

However, if batching is not enabled, the inputs are [6,4].

Are you suggesting I squeeze the inputs regardless?

Comment (tanmayv25) [reply inferentia/scripts/gen_triton_model.py:354]: Ohh.. I didn't see that before..  You should not design the python model on the basis of what perf analyzer generates. That is just a tool for running inferencing on server. Perf analyzer can be modified to generate requests expected by server but we must not loose generality.

So this is what your python model should be doing.. If the shape expected is in the following format: [batch size, X]. 
If the python model in its exec function gets three requests in the list as follows: [6, 4], [6,4] and [6, 4],
It should simply form a tensor of size [18, 4] and run that in neuron SDK. 

Now to test this you should run perf_analyzer as follows:

`perf_analyzer -m <model_name> --concurrency-range 10 -b 6` 

Note the model config should specify the dim as [4] and have max_batch_size specified as greater than or equal to 18. To form a batch of size 6, perf_analyzer goes through the tensors in a round-robin fashion.  Look at these documentation for more context.. https://github.com/triton-inference-server/server/blob/main/docs/model_configuration.md#inputs-and-outputs
https://github.com/triton-inference-server/server/blob/main/docs/perf_analyzer.md#real-input-data

**Review by tanmayv25 — APPROVED** [review]
